src/components/home/UserDisplay.tsx

```diff
@@ -1,16 +1,207 @@
+import ErrorCard from '../ErrorCard';
+import { isLenZero } from '../../utils';
 
-/**
- * Display the user's name.
- */
 const UserDisplay = () => {
+  /**
+   * Those are values used for setTimeouts, and the close animation.
+   */
+  const closeOnSuccessDelay = 500;
+  const closeDuration = 200;
+
+  /**
+   * State of the modal with it's setters.
+   */
+  const [isOpen, setIsOpen] = useState(false);
+
+  const closeModal = () => {
+    setIsOpen(false);
+  };
+
+  const openModal = () => {
+    setIsOpen(true);
+  };
+
+  /**
+   * The user in the context, the state of the input and
+   * a backup in case there is errors or edits without saving.
+   */
   const user = useContext(UserContext);
+  const [username, setUsername] = useState(user.get.name);
+  const [usernameBackup, setUsernameBackup] = useState(username);
+
+  /**
+   * Mutation to rename the user. The reset is used to clear
+   * the data and errors when the user closes the modal so
+   * it can be run again.
+   */
+  const {
+    data, isLoading, error, safeMutate, reset,
+  } = useRenameUser(username);
+
+  /**
+   * Change the state of the modal with some effects.
+   * Reset to clear errors after the close animation.
+   */
+  const exitModal = () => {
+    if (!data) setUsername(usernameBackup);
+    closeModal();
+
+    setTimeout(() => {
+      reset();
+    }, closeDuration);
+  };
+
+  /**
+   * Fire the mutation on save.
+   */
+  const saveModalOnClick = () => {
+    safeMutate();
+  };
+
+  /**
+   * Modify the user object as the user types.
+   */
+  useEffect(() => {
+    user.dispatch({
+      type: 'SET_NAME',
+      payload: { name: username },
+    });
+  }, [username]);
+
+  /**
+   * Update the backup and close the modal on success.
+   * Add a timeout with a reset to clear the data after
+   * the animation is done.
+   */
+  useEffect(() => {
+    setUsernameBackup(username);
+    closeModal();
+
+    setTimeout(() => {
+      reset();
+    }, closeOnSuccessDelay + closeDuration);
+  }, [data]);
+
+  /**
+   * Change the visuals of the modal based on
+   * the state of the server.
+   */
+  const saveButtonContent = () => {
+    if (isLoading) {
+      return <ClipLoader size="18px" color="white" />;
+    }
+
+    if (error) {
+      return 'Retry';
+    }
+
+    if (data) {
+      return <CheckIcon className="text-white w-6 h-6" />;
+    }
+
+    return 'Save';
+  };
+
+  const closeDurationClass = `duration-${closeDuration}`;
+
+  const closeDelayClass = data
+    ? `delay-${closeOnSuccessDelay}`
+    : '';
+
+  const saveButtonClass = data
+    ? 'bg-light-success-400 dark:bg-dark-success-500'
+    : '';
+
+  const errorMessage = () => {
+    if (isLenZero(username)) {
+      return (
+        <ErrorCard>
+          Empty names are not allowed.
+        </ErrorCard>
+      );
+    }
+
+    if (error) {
+      return (
+        <ErrorCard>
+          An error occurred while trying to rename you.
+        </ErrorCard>
```

Comment:
We will send friendly errors from the back. You can start printing the error if the code is 400 (bad request)